app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager/event_catcher/stream.rb

```diff
@@ -17,32 +15,21 @@ def stop
     @stop_polling = true
   end
 
-  def fake_events
-    [
-      OpenStruct.new(:name       => %w[instance_power_on instance_power_off].sample,
-                     :id         => Time.zone.now.to_i,
-                     :timestamp  => Time.zone.now,
-                     :vm_ems_ref => [1, 2].sample),
-      OpenStruct.new(:name       => %w[instance_power_on instance_power_off].sample,
-                     :id         => Time.zone.now.to_i + 1,
-                     :timestamp  => Time.zone.now,
-                     :vm_ems_ref => [1, 2].sample),
-    ]
-  end
-
   def poll
-    @ems.with_provider_connection do |_connection|
-      catch(:stop_polling) do
-        loop do
-          fake_events.each do |activity|
-            throw :stop_polling if @stop_polling
-            yield activity.to_h
-          end
-          sleep @poll_sleep
-        end
-      rescue => exception
-        raise ProviderUnreachable, exception.message
+    from_time = Time.now.utc.to_i
+
+    loop do
+      ems.with_provider_connection(:service => 'PCloudEventsApi') do |api|
+        # Could also use IBM Cloud Activity Tracker like the VPC provider
+        events = api.pcloud_events_getsince(ems.uid_ems, from_time).events
+        from_time = Time.now.utc.to_i
+
+        break if stop_polling
```

Comment:
I think you need this break outside of the `with_provider_connection` block for this to exit cleanly

Reply:
Also, the event catcher will finish processing events that are added to its internal queue before stopping (to prevent "in-flight" loss of events) so you probably want to yield the events to the caller then check for `stop_polling`

Reply:
LIke this 04bb108 ?

If so then should the VPC `EventCatcher::Stream` also switch to yield before `break if stop_polling`? https://github.com/ManageIQ/manageiq-providers-ibm_cloud/blob/master/app/models/manageiq/providers/ibm_cloud/vpc/cloud_manager/event_catcher/stream.rb#L35-L37

Reply:
Yes, and yes :wink:

Reply:
https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/304